File: Regexp/States.cs
using System;
using System.Text;

namespace Regexp
{
    public class States
    {
        static string printSpaces(int num){
            var sb = new StringBuilder();
            for (int i = 0; i < num; ++i)
            {
                sb.Append(" ");
            }
            return sb.ToString ();
        }
        public static string printnfa(NFA.State start, int depth)
        {
            var sb = new StringBuilder();
            if (depth > 100) {
                throw new Exception ("!!");
            }
            sb.Append(printSpaces(depth));
            sb.Append(String.Format("(c={0}, lastlist={1} \n",start.c, start.lastList));
            if (start.@out!=null){
                sb.Append(printSpaces(depth+1));
                sb.Append(",out=\n");
                sb.Append(printnfa(start.@out.Value, depth+2));
            }
            if (start.out1!=null){
                sb.Append(printSpaces(depth+1));
                sb.Append(",out1:\n");
                sb.Append(printnfa(start.out1.Value, depth+2));
            }
            sb.Append(printSpaces(depth));
            sb.Append(")\n");
            return sb.ToString ();
        }
    }
}

File: tests/Thompson.Regex.Tests/States.fs
module States
open System
open System.Text

    let rec printnfa (start:NFA.State<char>, depth:int)=
        if depth>100 then
            failwith "!!"
        else
            let sb = new StringBuilder()
            let printSpaces num=
                String.init num (fun _ -> " ")
            let append (v:string)=
                sb.Append(v) |> ignore

            append(printSpaces(depth));
            append(String.Format("(c={0}, id={1} \n", start.c, start.id));
            start.out |> Option.iter (fun out->
                append(printSpaces(depth+1))
                append(",out=\n")
                append(printnfa(out, depth+2))
            )
            start.out1 |> Option.iter (fun out->
                append(printSpaces(depth+1))
                append(",out1:\n")
                append(printnfa(out, depth+2))
            )
            append(printSpaces(depth))
            append(")\n")
            sb.ToString()
